### PR TITLE
NSG Updated After CLI Update to Instance_Config

### PR DESCRIPTION
**Pull request by nharper285**
## Summary of the Pull Request

_What is this about?_
This change updates all NSGs when a change is made to instance_config 

## PR Checklist
* [x] Applies to work item: #1320
* [x] CLA signed. If not, go over [here](https://cla.opensource.microsoft.com/microsoft/onefuzz) and sign the CLI.
* [ ] Tests added/passed
* [ ] Requires documentation to be updated
* [x] I've discussed this with core contributors already. If not checked, I'm ready to accept this work might be rejected in favor of a different grand plan. Issue number where discussion took place: #xxx

## Info on Pull Request

_What does this include?_
Changes to instance_config __init__.py. When a >onefuzz instance_config update @body.json is made, all NSGs are updated. 

## Validation Steps Performed

_How does someone test & validate?_

### CODE REVIEW

File: src/api-service/__app__/onefuzzlib/repro.py [hunk 1]
```diff
@@ -85,6 +86,23 @@ def init(self) -> None:
...
+                name=vm.region,
+                region=vm.region,
+            )
+            result = nsg.create()
+            if isinstance(result, Error):
+                self.set_failed(result)
+                return
+
+            nsg_config = NetworkSecurityGroupConfig()
+            nsg_config.allowed_ips = ["*"]
```
Comment (stishkin) [inline src/api-service/__app__/onefuzzlib/repro.py:99]: Do not forget to change this to set allowed IPs to values from Instance Config

File: src/api-service/__app__/onefuzzlib/azure/nsg.py [hunk 2]
```diff
@@ -0,0 +1,337 @@
...
+        # Will not exceed `max_rule_count` or max NSG priority (4096)
+        # due to earlier check of `len(all_sources)`.
+        priority += 1
+
+    nsg.security_rules = security_rules
+    return update_nsg(nsg)
+
+
+def clear_all_rules(name: str) -> Union[None, Error]:
+    return set_allowed(name, NetworkSecurityGroupConfig())
```
Comment (stishkin) [inline src/api-service/__app__/onefuzzlib/azure/nsg.py:175]: Can you add something like NetworkSecurityGroupConfig.empty()

So it is explicit that we are setting rules to an empty set

File: src/api-service/__app__/onefuzzlib/proxy.py [hunk 3]
```diff
@@ -90,6 +92,25 @@ def init(self) -> None:
...
+                region=self.region,
+            )
+
+            result = nsg.create()
+            if isinstance(result, Error):
+                self.set_failed(result)
+                return
+
+            nsg_config = NetworkSecurityGroupConfig()
+            nsg_config.allowed_ips = ["*"]
```
Comment (stishkin) [inline src/api-service/__app__/onefuzzlib/proxy.py:106]: Change allowed_ips from ["*"]  to the value set in InstanceConfig

File: src/api-service/__app__/onefuzzlib/azure/nsg.py (hunk 2 above)
Comment (stishkin) [reply src/api-service/__app__/onefuzzlib/azure/nsg.py:175]: yes

Comment (nharper285) [reply src/api-service/__app__/onefuzzlib/azure/nsg.py]: Okay I just updated. Let me know if this is what you want.

File: src/api-service/__app__/onefuzzlib/proxy.py (hunk 3 above)
Comment (nharper285) [reply src/api-service/__app__/onefuzzlib/proxy.py]: Changed.

File: src/api-service/__app__/instance_config/__init__.py [hunk 4]
```diff
@@ -30,8 +32,34 @@ def post(req: func.HttpRequest) -> func.HttpResponse:
             context="instance_config_update",
         )
 
+    update_nsg = False
+    if request.config.proxy_nsg_config and config.proxy_nsg_config:
+        request_config = request.config.proxy_nsg_config
```
[the remark this answers is not in the document]
Comment (nharper285) [reply src/api-service/__app__/instance_config/__init__.py:37]: This is already checked in a previous line: 
if not can_modify_config(req, config):
        return not_ok(
            Error(code=ErrorCode.INVALID_PERMISSION, errors=["unauthorized"]),
            context="instance_config_update",
        )

**Review by stishkin — APPROVED** [review]

File: src/api-service/__app__/onefuzzlib/repro.py (hunk 1 above)
Comment (nharper285) [reply src/api-service/__app__/onefuzzlib/repro.py]: Updated.

**Pull request merged**